onnxruntime/core/graph/graph.cc

```diff
@@ -1631,12 +1635,17 @@ common::Status Graph::TypeCheckInputsAndInitializers() {
         inferred_shape.add_dim()->set_dim_value(dim);
       }
       const TensorShapeProto* p_existing_shape = node_arg->Shape();
-      if (nullptr == p_existing_shape)
-        node_arg->SetShape(inferred_shape);
-      else {
+      if (nullptr == p_existing_shape) {
+        // use the inferred shape if this is a constant initializer (cannot be overridden).
+        // if not it has a matching graph input, and we prefer the shape info (or lack of info) from the graph input
+        if (graph_utils::IsConstantInitializer(*this, name, false)) {
+          node_arg->SetShape(inferred_shape);
```

Comment:
The logic looks fine. However, I wonder now whether there is some redundancy between this and the earlier logic when the NodeArgs are created (around new line 688)? Do we need to do this in both places, or is one of this redundant?

Reply:
As discussed, there's still scenarios where shape could be nullptr at this point and needs to be set.